### PR TITLE
feat: Add FileCopyMethod option / API

### PR DESCRIPTION
**Pull request by eth-p**
Implemented as discussed in #163 in [this comment](https://github.com/otiai10/copy/issues/163#issuecomment-2300899873).

A change was needed:

I have to check for the source file being deleted after the `FileCopyMethod` function is run.
   - The old function returned `nil` if the source was deleted.
   - When called from the refactored `fcopy`, it would think there was no error.
   - `fcopy` would then try to apply changes to the destination file that was never made, then error.

I also had to change the CI so I could run workflows on my own repo. You can drop that commit before merging it.

### CODE REVIEW

File: options.go [hunk 1]
```diff
@@ -27,6 +27,10 @@ type Options struct {
 	// RenameDestination can specify the destination file or dir name if needed to rename.
 	RenameDestination func(src, dest string) (string, error)
 
+	// FileCopyFunc specifies the method by which a regular file is copied.
+	// The default is CopyBytes.
+	FileCopyFunc FileCopyMethod
```
Comment (otiai10) [inline options.go:32]: What's the purpose of using different names here?
Both the name of prop and the type can be the same. Let me know your thoughts

File: copy_methods.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+import (
+	"errors"
+	"io"
+	"os"
+)
+
+// ErrUnsupportedCopyMethod is returned when the FileCopyMethod specified in
+// Options is not supported.
+var ErrUnsupportedCopyMethod = errors.New(
```
Comment (otiai10) [inline copy_methods.go:11]: What's the case when we use it?

File: copy.go [hunk 3]
```diff
@@ -86,66 +87,38 @@ func copyNextOrSkip(src, dest string, info os.FileInfo, opt Options) error {
...
 	if err = os.MkdirAll(filepath.Dir(dest), os.ModePerm); err != nil {
 		return
 	}
 
-	f, err := os.Create(dest)
-	if err != nil {
-		return
+	// Use FileCopyMethod to do copy.
+	if err = opt.FileCopyFunc.fcopy(src, dest, info, opt); err != nil {
+		if pathError, ok := err.(*os.PathError); ok && errors.Is(err, os.ErrNotExist) && pathError.Path == src {
```
Comment (otiai10) [inline copy.go:96]: Just at a glance, it looks like increasing complexity of source code.
Do we really have a specific issue we need to solve in this PR?
or do you think we can separate PR for different point of interest?

File: options.go (hunk 1 above)
Comment (eth-p) [reply options.go:32]: Good point. I'll make them both `FileCopyMethod`

File: copy_methods.go (hunk 2 above)
Comment (eth-p) [reply copy_methods.go:11]: So the user can tell if the copy failed because of no support, instead of another error like permission:

```go
err := copy.Copy("/from", "/to", copy.Options {
    FileCopyMethod: copy.ReflinkCopy,
})

if errors.Is(err, copy.ErrUnsupportedCopyMethod) {
    // Retry with other method
    err = copy.Copy("/from", "/to", copy.Options {
        FileCopyMethod: copy.CopyBytes,
    })
}
```

File: copy.go (hunk 3 above)
Comment (eth-p) [reply copy.go:96]: If I don't do this here, it breaks the `file is deleted while copying` test. Let me see if I can do something else though

Comment (eth-p) [reply copy.go]: I changed it to have a `skipFile` return instead

File: copy_methods.go (hunk 2 above)
Comment (otiai10) [reply copy_methods.go:11]: quick question: Do we use this error in this current version?

**Review by otiai10 — COMMENTED** [review]
Overall, it looks good to me. I left a small question. Please check it out

File: copy_methods.go (hunk 2 above)
Comment (eth-p) [reply copy_methods.go:11]: Not yet. It's used in a later pull request I made

Comment (otiai10) [reply copy_methods.go:11]: OK, got it. In this case I'll merge it.

Very personally speaking, I would not like to mix what we don't use at that moment. Similar idea to YAGNI.

Comment (eth-p) [reply copy_methods.go:11]: I can remove it and add it back in the other PR, if that's preferred?

Comment (otiai10) [reply copy_methods.go:11]: No need. Thanks for your consideration. I understand why it's here and we don't have any doubt we will need it ;)

**Review by otiai10 — APPROVED** [review]
